### PR TITLE
MueLu: Reduce communication in setup

### PR DESCRIPTION
**Pull request by cgcgcg**
@trilinos/tpetra @trilinos/ifpack2 @trilinos/muelu 

## Motivation
Reduce communication during MueLu setup by making global constants optional.
- Fix logic in `Aggregates` for global reduction.
- Adjust verbosity level "Medium" to skip "Statistics1". This allows to run with timings but without global stats.
- Compute global statistics for matrices only with "Statistics1".

### CODE REVIEW

File: packages/muelu/src/MueCentral/MueLu_Hierarchy_def.hpp [hunk 1]
```diff
@@ -1241,8 +1246,9 @@ void Hierarchy<Scalar, LocalOrdinal, GlobalOrdinal, Node>::describe(Teuchos::Fan
     std::vector<Xpetra::global_size_t> nnzPerLevel;
     std::vector<Xpetra::global_size_t> rowsPerLevel;
     std::vector<int> numProcsPerLevel;
-    bool someOpsNotMatrices             = false;
-    const Xpetra::global_size_t INVALID = Teuchos::OrdinalTraits<Xpetra::global_size_t>::invalid();
+    bool someOpsNotMatrices                 = false;
+    const Xpetra::global_size_t OPERATOR    = -10;
+    const Xpetra::global_size_t UNAVAILABLE = -9;
     for (int i = 0; i < numLevels; i++) {
```
Comment (csiefer2) [inline packages/muelu/src/MueCentral/MueLu_Hierarchy_def.hpp:1252]: global_size_t is actually a size_t which is unsigned

Comment (csiefer2) [reply packages/muelu/src/MueCentral/MueLu_Hierarchy_def.hpp:1252]: Maybe `OrdinalTraits::invalid()-1`?

File: packages/tpetra/core/src/Tpetra_Map_def.hpp [hunk 2]
```diff
@@ -757,10 +757,17 @@ void Map<LocalOrdinal, GlobalOrdinal, Node>::
...
-    distributed_ = params->get("distributed", true) && (comm->getSize() > 1);
+  else {
+    if (comm->getSize() == 1) {
+      minAllGID_           = minMyGID_;
+      maxAllGID_           = maxMyGID_;
+      distributed_         = false;
+      haveGlobalConstants_ = true;
+    } else
+      distributed_ = params->get("distributed", true);
+  }
```
Comment (csiefer2) [inline packages/tpetra/core/src/Tpetra_Map_def.hpp:766]: Can we have `computeGlobalConstants()` handle this instead?

File: packages/tpetra/core/src/Tpetra_Map_def.hpp [hunk 3]
```diff
@@ -1020,10 +1027,17 @@ Map<LocalOrdinal, GlobalOrdinal, Node>::
...
```
Comment (csiefer2) [inline packages/tpetra/core/src/Tpetra_Map_def.hpp:1031]: See above

File: packages/muelu/src/MueCentral/MueLu_Hierarchy_def.hpp (hunk 1 above)
Comment (cgcgcg) [reply packages/muelu/src/MueCentral/MueLu_Hierarchy_def.hpp:1252]: done

File: packages/tpetra/core/src/Tpetra_Map_def.hpp (hunk 2 above)
Comment (cgcgcg) [reply packages/tpetra/core/src/Tpetra_Map_def.hpp:766]: done.

File: packages/muelu/src/MueCentral/MueLu_Hierarchy_def.hpp [hunk 4]
```diff
@@ -1241,8 +1246,9 @@ void Hierarchy<Scalar, LocalOrdinal, GlobalOrdinal, Node>::describe(Teuchos::Fan
     std::vector<Xpetra::global_size_t> nnzPerLevel;
     std::vector<Xpetra::global_size_t> rowsPerLevel;
     std::vector<int> numProcsPerLevel;
-    bool someOpsNotMatrices             = false;
-    const Xpetra::global_size_t INVALID = Teuchos::OrdinalTraits<Xpetra::global_size_t>::invalid();
+    bool someOpsNotMatrices                 = false;
+    const Xpetra::global_size_t OPERATOR    = Teuchos::OrdinalTraits<Xpetra::global_size_t>::invalid();
+    const Xpetra::global_size_t UNAVAILABLE = Teuchos::OrdinalTraits<Xpetra::global_size_t>::invalid() - 1;
```
Comment (csiefer2) [inline packages/muelu/src/MueCentral/MueLu_Hierarchy_def.hpp]: Hmmmm.  Are we sure this is positive?

File: packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp [hunk 5]
```diff
@@ -7515,26 +7513,9 @@ void CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
...
-  int reduced_mismatch = 0;
-  if (isMM && !overrideAllreduce) {
-    // Test for pathological matrix transfer
-    const bool source_vals = !getGraph()->getImporter().is_null();
-    const bool target_vals = !(rowTransfer.getExportLIDs().size() == 0 ||
-                               rowTransfer.getRemoteLIDs().size() == 0);
-    mismatch               = (source_vals != target_vals) ? 1 : 0;
-    iallreduceRequest =
-        ::Tpetra::Details::iallreduce(mismatch, reduced_mismatch,
-                                      Teuchos::REDUCE_MAX, *(getComm()));
```
Comment (csiefer2) [inline packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp:7541]: Should we test this "pathological" corner case?

**Review by csiefer2 — CHANGES_REQUESTED** [review]

File: packages/muelu/src/MueCentral/MueLu_Hierarchy_def.hpp (hunk 4 above)
Comment (cgcgcg) [reply packages/muelu/src/MueCentral/MueLu_Hierarchy_def.hpp]: I have replaced `invalid-1` with `max`. The two are different.

File: packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp (hunk 5 above)
Comment (cgcgcg) [reply packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp:7541]: I have added testing of the matrix-matrix tests with optimized TAFC. I verified that the pathological case does get triggered by several of the test cases. The global reduction does not seem to be required for correctness.

File: packages/muelu/src/MueCentral/MueLu_Hierarchy_def.hpp [hunk 6]
```diff
@@ -1241,8 +1246,9 @@ void Hierarchy<Scalar, LocalOrdinal, GlobalOrdinal, Node>::describe(Teuchos::Fan
     std::vector<Xpetra::global_size_t> nnzPerLevel;
     std::vector<Xpetra::global_size_t> rowsPerLevel;
     std::vector<int> numProcsPerLevel;
-    bool someOpsNotMatrices             = false;
-    const Xpetra::global_size_t INVALID = Teuchos::OrdinalTraits<Xpetra::global_size_t>::invalid();
+    bool someOpsNotMatrices                 = false;
+    const Xpetra::global_size_t OPERATOR    = Teuchos::OrdinalTraits<Xpetra::global_size_t>::invalid();
+    const Xpetra::global_size_t UNAVAILABLE = Teuchos::OrdinalTraits<Xpetra::global_size_t>::max();
     for (int i = 0; i < numLevels; i++) {
```
Comment (csiefer2) [inline packages/muelu/src/MueCentral/MueLu_Hierarchy_def.hpp:1252]: Maybe consider adding a test to Teuchos to ensure max() != invalid()

**Review by csiefer2 — APPROVED** [review]

File: packages/muelu/src/MueCentral/MueLu_Hierarchy_def.hpp (hunk 6 above)
Comment (cgcgcg) [reply packages/muelu/src/MueCentral/MueLu_Hierarchy_def.hpp:1252]: https://github.com/trilinos/Trilinos/blob/12801899faeb8e7258720b7c057ed280eb310f5d/packages/teuchos/core/test/ScalarTraits/ScalarTraits_test.cpp#L235-L241
